File: frontend/src/render/canonical.ts
// Canonicalization: raw answer (what was entered) → value (canonical) — §2.3, §2.8.
//
// Two rules are exercised by the gallery:
//   1. allow_other:  raw "other:capybara" → value "capybara" (radio),
//      and element-wise for checkbox arrays.
//   2. panel_order variant: a positional choice (Left/Tie/Right) canonicalizes
//      to the item chosen (A/Tie/B), using the slot's variant so bias is
//      measurable and Left/Right identity never leaks to the annotator (§2.7, §4).
//
// value === raw for variant-free templates with no Other selection.

import type { InputField, TemplateSchema } from "../api/types";

export const OTHER_PREFIX = "other:";

function stripOther(v: unknown): unknown {
  if (typeof v === "string" && v.startsWith(OTHER_PREFIX)) {
    return v.slice(OTHER_PREFIX.length);
  }
  return v;
}

// Map a positional label to the canonical item under a variant string.
// "Left"/"Right" → variant[0]/variant[1]; "Tie" and anything else pass through.
function canonicalizePositional(label: unknown, variant: string | null): unknown {
  if (typeof label !== "string" || !variant) return label;
  const l = label.trim().toLowerCase();
  if (l === "left") return variant[0] ?? label;
  if (l === "right") return variant[variant.length - 1] ?? label;
  return label; // tie, center, etc.
}

// --- M6 value-shape canonicalizers (§2.3) ----------------------------------
// Mirrors backend `services/quality/canonical.py`. The server recomputes all of
// this and its answer is the one that's stored (§2.6) — keeping the two in step
// is what lets a client-side preview show the value that will actually be saved.

const TRUE_TOKENS = new Set(["true", "yes", "y", "1", "on"]);
const FALSE_TOKENS = new Set(["false", "no", "n", "0", "off"]);

function toBool(raw: unknown): unknown {
  if (typeof raw === "boolean") return raw;
  if (typeof raw === "number") return raw !== 0;
  if (typeof raw === "string") {
    const token = raw.trim().toLowerCase();
    if (TRUE_TOKENS.has(token)) return true;
    if (FALSE_TOKENS.has(token)) return false;
  }
  return raw;
}

function toNumber(raw: unknown, integral = false): unknown {
  if (typeof raw === "boolean") return raw;
  if (typeof raw === "number") return integral ? Math.trunc(raw) : raw;
  if (typeof raw === "string" && raw.trim()) {
    const n = Number(raw.trim());
    if (!Number.isFinite(n)) return raw;
    return integral || Number.isInteger(n) ? Math.trunc(n) : n;
  }
  return raw;
}

// Trim, lower-case and de-duplicate — without folding, "Spam", "spam " and
// "spam" are three different labels and every agreement metric is noise.
function toTags(raw: unknown): unknown {
  const items = typeof raw === "string" ? raw.split(",") : Array.isArray(raw) ? raw : null;
  if (items === null) return raw;
  const out: string[] = [];
  for (const item of items) {
    if (typeof item !== "string") continue;
    const tag = item.trim().toLowerCase().split(/\s+/).join(" ");
    if (tag && !out.includes(tag)) out.push(tag);
  }
  return out;
}

// `ranking` — an ordered array; order is the answer, so nothing is sorted.
function toStringList(raw: unknown): unknown {
  return Array.isArray(raw) ? raw.filter((v): v is string => typeof v === "string") : raw;
}

const BY_TYPE: Partial<Record<string, (raw: unknown) => unknown>> = {
  boolean: toBool,
  number: (raw) => toNumber(raw),
  slider: (raw) => toNumber(raw),
  rating: (raw) => toNumber(raw, true),
  likert: (raw) => toNumber(raw, true),
  tags: toTags,
  ranking: toStringList,
};

export function canonicalizeInput(
  input: InputField,
  raw: unknown,
  opts: { positionalVariant: string | null },
): unknown {
  const { positionalVariant } = opts;

  // choice_buttons under a panel_order variant → item canonicalization.
  if (input.type === "choice_buttons" && positionalVariant) {
    return canonicalizePositional(raw, positionalVariant);
  }

  // allow_other → strip the "other:" prefix.
  if (input.allow_other) {
    if (Array.isArray(raw)) return raw.map(stripOther);
    return stripOther(raw);
  }

  const converter = BY_TYPE[input.type];
  if (converter) return converter(raw);

  return raw;
}

export function canonicalize(
  schema: TemplateSchema,
  raw: Record<string, unknown>,
  positionalVariant: string | null,
): Record<string, unknown> {
  const value: Record<string, unknown> = {};
  const byId = new Map(schema.inputs.map((i) => [i.id, i]));
  for (const [id, rawVal] of Object.entries(raw)) {
    const input = byId.get(id);
    value[id] = input
      ? canonicalizeInput(input, rawVal, { positionalVariant })
      : rawVal;
  }
  return value;
}

File: backend/app/services/quality/canonical.py
"""Server-side canonicalization: raw answer → canonical value (§2.3, §2.6, §2.8).

M3 canonicalized in the browser and posted both ``raw`` and ``value``. From M4 the
backend is authoritative: gold grading, agreement and merge all read ``value``, so a
client that computes it wrongly (or maliciously) would corrupt every quality signal
downstream. The client still sends ``value`` — it is now advisory and only used for
input types this module declines to canonicalize.

Mirrors ``frontend/src/render/canonical.ts``; the two are kept in sync by the
shared gallery fixtures (a template that canonicalizes differently on the two
sides shows up as a gold/agreement discrepancy in tests).
"""

from __future__ import annotations

from typing import Any

OTHER_PREFIX = "other:"

# Positional option labels that a variant maps onto items (§2.7).
_LEFT_LABELS = {"left", "a", "first"}
_RIGHT_LABELS = {"right", "b", "second"}


def _strip_other(value: Any) -> Any:
    """``"other:capybara"`` → ``"capybara"`` (the escape hatch, §2.3)."""
    if isinstance(value, str) and value.startswith(OTHER_PREFIX):
        return value[len(OTHER_PREFIX) :]
    return value


def positional_variant(
    template_schema: dict[str, Any], variant: dict[str, Any] | None
) -> str | None:
    """The variant string (e.g. ``"BA"``) when the template declares one, else None."""
    variants = template_schema.get("variants")
    if not variants or not variant:
        return None
    dimension = variants.get("dimension")
    if not dimension:
        return None
    value = variant.get(dimension)
    return value if isinstance(value, str) else None


def canonicalize_positional(label: Any, variant: str | None) -> Any:
    """Map a side clicked onto the item shown there.

    Under ``panel_order="BA"`` the left panel holds item B, so a "Left" click
    canonicalizes to ``"B"``. Ties and unrecognized labels pass through. This is
    what makes position bias measurable (§9): ``raw`` keeps the side, ``value``
    keeps the item.
    """
    if not isinstance(label, str) or not variant:
        return label
    key = label.strip().lower()
    if key in _LEFT_LABELS:
        return variant[0]
    if key in _RIGHT_LABELS:
        return variant[-1]
    return label


# --- M6 value-shape canonicalizers (§2.3, §2.6 step 3) ----------------------
#
# Each new input type declares a value shape; these functions are the "optional
# canonicalizer" of the extensibility contract. They are deliberately forgiving
# about what arrives (an HTML number field yields a string, a checkbox yields
# "true") and strict about what is stored, because gold grading and agreement
# only ever read the canonical side.

_TRUE_TOKENS = {"true", "yes", "y", "1", "on"}
_FALSE_TOKENS = {"false", "no", "n", "0", "off"}


def _to_bool(raw: Any) -> Any:
    if isinstance(raw, bool):
        return raw
    if isinstance(raw, int | float):
        return bool(raw)
    if isinstance(raw, str):
        token = raw.strip().lower()
        if token in _TRUE_TOKENS:
            return True
        if token in _FALSE_TOKENS:
            return False
    return raw  # unrecognized → pass through; a gold will simply not match


def _to_number(raw: Any, *, integral: bool = False) -> Any:
    if isinstance(raw, bool):
        return raw
    if isinstance(raw, int | float):
        return int(raw) if integral else raw
    if isinstance(raw, str) and raw.strip():
        try:
            number = float(raw.strip())
        except ValueError:
            return raw
        if integral or number.is_integer():
            return int(number)
        return number
    return raw


def _to_tags(raw: Any) -> Any:
    """Free-form tag entry → a de-duplicated, trimmed, lower-cased string array.

    Case- and whitespace-folding is what makes tags comparable at all: without
    it "Spam", "spam " and "spam" are three different labels and every agreement
    metric over the key is noise.
    """
    items: list[Any]
    if isinstance(raw, str):
        items = raw.split(",")
    elif isinstance(raw, list | tuple):
        items = list(raw)
    else:
        return raw
    out: list[str] = []
    for item in items:
        if not isinstance(item, str):
            continue
        tag = " ".join(item.strip().lower().split())
        if tag and tag not in out:
            out.append(tag)
    return out


def _to_string_list(raw: Any) -> Any:
    """``ranking`` — an ordered array; order is the answer, so nothing is sorted."""
    if isinstance(raw, list | tuple):
        return [v for v in raw if isinstance(v, str)]
    return raw


_BY_TYPE = {
    "boolean": _to_bool,
    "number": _to_number,
    "slider": _to_number,
    "rating": lambda raw: _to_number(raw, integral=True),
    "likert": lambda raw: _to_number(raw, integral=True),
    "tags": _to_tags,
    "ranking": _to_string_list,
}


def canonicalize_input(field: dict[str, Any], raw: Any, *, variant_str: str | None) -> Any:
    """Canonicalize one input's raw answer."""
    itype = field.get("type")
    if itype == "choice_buttons" and variant_str:
        return canonicalize_positional(raw, variant_str)
    if field.get("allow_other"):
        if isinstance(raw, list):
            return [_strip_other(v) for v in raw]
        return _strip_other(raw)
    converter = _BY_TYPE.get(itype)
    if converter is not None:
        return converter(raw)
    return raw


def canonicalize(
    template_schema: dict[str, Any],
    raw: dict[str, Any],
    variant: dict[str, Any] | None,
) -> dict[str, Any]:
    """Canonicalize a whole submission. Unknown keys pass through untouched."""
    variant_str = positional_variant(template_schema, variant)
    by_id = {f["id"]: f for f in template_schema.get("inputs", []) if "id" in f}
    out: dict[str, Any] = {}
    for key, value in raw.items():
        field = by_id.get(key)
        out[key] = canonicalize_input(field, value, variant_str=variant_str) if field else value
    return out
